icu4c/source/common/locid.cpp

```diff
@@ -1559,11 +1646,40 @@ AliasReplacer::replace(const Locale& locale, CharString& out, UErrorCode status)
     if (U_FAILURE(status)) { return false; }
     // Nothing changed and we know the order of the vaiants are not change
     // because we have no variant or only one.
-    if (changed == 0 && variants.size() <= 1) {
+    const char* extensionsStr = locale_getKeywordsStart(locale.getName());
+    if (changed == 0 && variants.size() <= 1 && extensionsStr == nullptr) {
         return false;
     }
     outputToString(out, status);
+    if (U_FAILURE(status)) {
+        return false;
+    }
     if (extensionsStr != nullptr) {
+        changed = 0;
+        Locale temp(locale);
+        LocalPointer<icu::StringEnumeration> iter(locale.createKeywords(status));
+        if (U_SUCCESS(status) && !iter.isNull()) {
+            const char* key;
+            while ((key = iter->next(nullptr, status)) != nullptr) {
+                if (uprv_strcmp("sd", key) == 0 || uprv_strcmp("rg", key) == 0) {
+                    status = U_ZERO_ERROR;
+                    CharString value;
+                    CharStringByteSink valueSink(&value);
+                    locale.getKeywordValue(key, valueSink, status);
+                    if (U_FAILURE(status)) {
+                        status = U_ZERO_ERROR;
+                        continue;
+                    }
+                    if (replaceSubdivision(value, status)) {
+                        changed++;
+                    }
+                    temp.setKeywordValue(key, value.data(), status);
```

Comment:
Sorry, found a problem: if U_FAILURE after this line, then the loop just continues. Worst case, you get to the second relevant key and clobber the status back to a zero-error.

Please bail out of the loop after this line if there is a failure code. I am not looking at enough context to see if you just need to "break" or if you should return immediately with something.